File: hub/homepilot/core/pushziel.py
"""Wohin ein Tipp auf eine Nachricht führt.

Die App wusste das lange für vier Fälle: Kamera, Batterie, Klingel,
Alarm. Alle anderen fünfundzwanzig Meldungen öffneten sie auf der
Startseite - «Milch steht auf der Einkaufsliste» genauso wie «Wasser im
Keller». Man tippt, weil man etwas tun will, und landet dort, wo man
ohnehin gelandet wäre.

Die Regel dafür gehört hierher und nicht in die App: Hier weiss man, was
für eine Meldung das ist. Die App muss nur wissen, wie man zu einem Raum
kommt (lib/pushziel.ts).

Die Schreibweise ist absichtlich schlicht - ``art`` oder ``art:wert``,
eine Zeichenkette, die durch JSON und durch die Push-Dienste geht, ohne
unterwegs etwas zu verlieren.
"""

from __future__ import annotations

## Je Kategorie der Ort, an dem man etwas tun kann.
#
# Nicht «wo steht davon etwas», sondern «wo tut man etwas»: Der
# Wasseralarm gehört nicht in eine Liste vergangener Meldungen, sondern
# in den Raum, in dem der Melder hängt.
ZIELE: dict[str, str] = {
    # Sicherheit: die Anlage selbst, dort steht auch der Verlauf.
    "alarm": "bereich:alarm",
    "alarm_arming": "bereich:alarm",
    # Betrieb: Integrationen und Platte stehen im System-Bereich.
    "outage": "bereich:system",
    "disk": "bereich:system",
    # Was mit einem Gerät nicht stimmt, steht auf einem Blatt zusammen -
    # samt Quittieren.
    "device_down": "sorgen",
    "maintenance": "sorgen",
    "battery": "batterien",
    # Wetter und Haushalt: die Startseite trägt die Karten dazu.
    "frost": "start",
    "rain": "start",
    "plants": "start",
    "appliance": "start",
    "morning": "start",
    "test": "start",
    "open": "offen",
    "leak": "sorgen",
    "doorbell": "klingel",
    "timer": "timer",
    # Familie: je Meldung ihre Kachel.
    "tasks": "familie:tasks",
    "shopping": "familie:shopping",
    "calendar": "familie:kalender",
    "medication": "familie:medications",
    "birthday": "familie:kalender",
    "weekahead": "familie:woche",
    # Wer wo ist, steht unter «Familie und Freunde».
    "presence": "bereich:personen",
    # Bewegung sieht man auf der Kamera - welche, sagt die Nachricht
    # selbst (data.camera).
    "camera_motion": "start",
}

## Meldungen, bei denen das Gerät selbst der beste Ort ist.
#
# Ein offenes Fenster ist im Raum zu schliessen, nicht in einer Liste;
# ein Wassermelder schickt einen dorthin, wo man den Hahn zudreht. Bei
# den anderen (Batterie, ausgefallenes Gerät) ist der Sammelplatz
# besser: Dort steht der Knopf zum Quittieren.
AM_GERAET = frozenset({"open", "leak", "appliance"})


def ziel_fuer(category: str | None, entity_id: str | None = None) -> str | None:
    """Das Ziel zu einer Kategorie (rein, testbar).

    ``None`` heisst: kein besonderer Ort - dann öffnet die App sich
    einfach, wie bisher. Das gilt auch für Meldungen aus Abläufen: Dort
    entscheidet der Ablauf selbst, wohin es geht.
    """
    if not category:
        return None
    if entity_id and category in AM_GERAET:
        return f"geraet:{entity_id}"
    return ZIELE.get(category)

File: app/src/lib/pushziel.ts
/**
 * Wohin ein Tipp auf eine Nachricht führt.
 *
 * Bisher wusste die App das für vier Fälle: Kamera, Batterie, Klingel,
 * Alarm. Alle anderen fünfundzwanzig Meldungen öffneten die App auf der
 * Startseite - «Milch steht auf der Einkaufsliste» genauso wie «Wasser
 * im Keller». Man tippt, weil man etwas tun will, und landet dort, wo
 * man ohnehin gelandet wäre.
 *
 * Statt für jede Meldung eine Regel in der App zu schreiben, sagt jede
 * Nachricht selbst, wohin sie gehört: ein Feld `ziel` in den Daten, das
 * der Hub setzt (core/watchdog.py, core/automation.py). Die App muss
 * dafür nichts über Wassermelder wissen - nur, wie man zu einem Raum
 * kommt.
 *
 * Die Schreibweise ist absichtlich schlicht: `art` oder `art:wert`, eine
 * Zeichenkette, die durch JSON und durch die Push-Dienste geht, ohne
 * unterwegs etwas zu verlieren.
 */
import { Section } from '../components/Rail';

export type Ziel =
  /** Die Startseite – wenn nichts Besseres gemeint ist. */
  | { art: 'start' }
  /** Ein Bereich der App: 'system', 'alarm', 'energy', 'personen' … */
  | { art: 'bereich'; bereich: Section }
  /** Ein Raum in der Raumansicht. */
  | { art: 'raum'; raum: string }
  /** Eine Kamera im Vollbild. */
  | { art: 'kamera'; entityId: string }
  /** Ein einzelnes Gerät – die App springt in seinen Raum. */
  | { art: 'geraet'; entityId: string }
  /** Eine Kachel im Familienbereich: 'shopping', 'tasks', 'kalender' … */
  | { art: 'familie'; modul: string }
  /** Die Geräteliste mit aufgeklappten Batterien. */
  | { art: 'batterien' }
  /** Das Blatt «Nicht in Ordnung». */
  | { art: 'sorgen' }
  /** Der Küchen-Timer. */
  | { art: 'timer' }
  /** Das Klingel-Vollbild. */
  | { art: 'klingel'; entityId?: string }
  /** Die Liste der offenen Fenster und Türen. */
  | { art: 'offen' };

/** Bereiche, in die eine Nachricht führen darf. Absichtlich eine Liste
 *  und keine offene Zeichenkette: Was der Hub schickt, kommt aus einer
 *  älteren Fassung, sobald jemand ein Update aufschiebt. */
const BEREICHE: Section[] = [
  'start',
  'home',
  'light',
  'covers',
  'cameras',
  'family',
  'devices',
  'automations',
  'system',
  'energy',
  'alarm',
  'speakers',
  'users',
  'personen',
  'activity',
  'widgets',
  'account',
  'connection',
];

/** Kacheln des Familienbereichs (screens/family/bausteine.tsx). */
const MODULE = [
  'kalender',
  'tasks',
  'shopping',
  'meals',
  'pins',
  'rewards',
  'contacts',
  'routines',
  'packlists',
  'countdowns',
  'reminders',
  'recipes',
  'documents',
  'chores',
  'woche',
  'emergency',
  'medications',
  'babysitter',
  'members',
];

/**
 * Aus dem, was an der Nachricht hängt, ein Ziel machen (rein, testbar).
 *
 * `null` heisst: Es ist nichts Sinnvolles dabei - dann bleibt es beim
 * blossen Öffnen der App, wie bisher.
 *
 * Die alten Felder (`camera`, `type`) werden weiter verstanden: Zwischen
 * einem neuen Hub und einer alten App - oder umgekehrt - liegen bei uns
 * regelmässig ein paar Tage.
 */
export function zielAus(data: {
  ziel?: unknown;
  camera?: unknown;
  entityId?: unknown;
  type?: unknown;
}): Ziel | null {
  const roh = typeof data.ziel === 'string' ? data.ziel.trim() : '';
  const entityId = typeof data.entityId === 'string' ? data.entityId : undefined;
  if (roh) {
    const geparst = ausSchluessel(roh, entityId);
    if (geparst) return geparst;
  }
  // Der alte Weg, unverändert.
  if (typeof data.camera === 'string' && data.camera) {
    return { art: 'kamera', entityId: data.camera };
  }
  if (data.type === 'doorbell') return { art: 'klingel', entityId };
  if (data.type === 'battery') return { art: 'batterien' };
  if (data.type === 'alarm') return { art: 'bereich', bereich: 'alarm' };
  return null;
}

function ausSchluessel(roh: string, entityId?: string): Ziel | null {
  const [art, ...rest] = roh.split(':');
  const wert = rest.join(':').trim();
  switch (art) {
    case 'start':
      return { art: 'start' };
    case 'batterien':
      return { art: 'batterien' };
    case 'sorgen':
      return { art: 'sorgen' };
    case 'timer':
      return { art: 'timer' };
    case 'offen':
      return { art: 'offen' };
    case 'klingel':
      return { art: 'klingel', entityId: wert || entityId };
    case 'bereich':
      return BEREICHE.includes(wert as Section)
        ? { art: 'bereich', bereich: wert as Section }
        : null;
    case 'raum':
      return wert ? { art: 'raum', raum: wert } : null;
    case 'kamera':
      return wert || entityId
        ? { art: 'kamera', entityId: (wert || entityId) as string }
        : null;
    case 'geraet':
      return wert || entityId
        ? { art: 'geraet', entityId: (wert || entityId) as string }
        : null;
    case 'familie':
      return MODULE.includes(wert) ? { art: 'familie', modul: wert } : null;
    default:
      return null;
  }
}
